Data scouting classes and producers (74X)

duanders opened the pull request:
Backport of https://github.com/cms-sw/cmssw/pull/10258/commits to 74X.

Commit:
Implement data formats and producers for scouting with jets

cmsbuild commented:
A new Pull Request was created by @duanders (Dustin Anderson) for CMSSW_7_4_X.

Data scouting classes and producers (74X)

It involves the following packages:

DataFormats/HLTReco
HLTrigger/JetMET


@Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

Martin-Grunewald commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e1e612bbcf7a6d59d04cf181f775e4e7d203eede
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/6769/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/6769/summary.html

Martin-Grunewald commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

cmsbuild commented:
+1
Tested at: d20e5f8cb567cb578240ee7fb62e1c3358dba78d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/6818/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/6818/summary.html

Martin-Grunewald commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_4_X IBs once checked with relvals in the development release cycle of CMSSW (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

Martin-Grunewald commented:
@davidlange6 
This is urgent for scouting triggers - please integrate as soon as possible for 748 - 
(75X #10377 and 76X #10258 PRs exist as well).
Thanks
Martin

duanders commented:
Hi,
Can we move forward with the integration of this PR and the others?  Is any other test necessary or is everything ok?

Dr15Jones commented:
I was asked where I thought these should live. I'd vote for DataFormats/Scouting if we are talking about new types.

duanders commented:
Hi @Dr15Jones @Martin-Grunewald ,
This is where they were originally.  Should I move them back?

Martin-Grunewald commented:
Hmm, we are moving in circles. A new subsystem/package for a couple of classes?

fwyzard commented:
Apologies for checking this so late, I've been mostly offline the past two
weeks.

Could you delay the integration till Tuesday morning, CERN time ?

Ciao,
.Andrea

Dr15Jones commented:
The issue is I've been told that these classes are going to be Core's responsibility. Therefore we will need to sign off on any change. If they are in their own package the seperation of responsibility is clear. If they are DataFormats/HLTReco then Core will have to sign off on changes to any of the other classes in that package.

davidlange6 commented:
which is to say these will be part of the raw data (for some streams). I agree with the plan for a separate package. 

> On Aug 10, 2015, at 3:43 PM, Chris Jones <notifications@github.com> wrote:
> 
> The issue is I've been told that these classes are going to be Core's responsibility. Therefore we will need to sign off on any change. If they are in their own package the seperation of responsibility is clear. If they are DataFormats/HLTReco then Core will have to sign off on changes to any of the other classes in that package.
> 
> —
> Reply to this email directly or view it on GitHub.
>

Martin-Grunewald commented:
Hmm well fine with me...

duanders commented:
Ok.  I will move the classes to DataFormats/Scouting, and remove "HLT" from the class names.  

Just to confirm, the producers should still live in HLTrigger/JetMET?

Dr15Jones commented:
The producers are fine where they are.

fwyzard commented:
Dustin, can you explain what is the purpose of
```
        float userFloat_;
        std::vector<int> constituents_;
```

in the `HLTScoutingPFJet` class ?

.A

duanders commented:
@fwyzard 
The `userFloat_` is to hold any extra ID variable that we choose.  Right now the producer sets this to 0 for all jets, but it leaves room to add something in the future.  For calo jets we are currently working on an online MVA to discriminate pileup, and if it is implemented for scouting then we will use this variable to store the discriminant value.  
The `constituents_` vector contains the indices of the PF candidates (as HLTScoutingParticles) associated with the jet.

fwyzard commented:
-1

> The userFloat_ is to hold any extra ID variable that we choose. Right now the producer sets this to 0 for all jets, but it leaves room to add something in the future. For calo jets we are currently working on an online MVA to discriminate pileup, and if it is implemented for scouting then we will use this variable to store the discriminant value.

sorry, but I'm strongly against storing a value without any explicit definition of its meaning.
With your current approach there is no way to track the meaning of the `userFloat_` value, and there is no guarantee that it will not change meaning from one run to an other.

Rather, please add an appropriately named member variable.


> The constituents_ vector contains the indices of the PF candidates (as HLTScoutingParticles) associated with the jet. 

Shouldn't you use an edm::RefVector for this, to keep track of the actual collection of HLTScoutingParticles ?

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

duanders commented:
@fwyzard 
Both of your comments make sense.  I'll make the changes tomorrow.

duanders commented:
I think there is a technical issue with the edm::RefVector approach, namely that the ScoutingPFJets and the ScoutingParticles are produced in the same module, making it impossible for one collection to contain refs to the other.  Is this correct, and do I need to create a separate producer for ScoutingParticles in this approach?

davidlange6 commented:
In any case I do not think we want refvectors in the raw data.



On Aug 11, 2015, at 1:21 PM, Dustin Anderson <notifications@github.com<mailto:notifications@github.com>> wrote:


I think there is a technical issue with the edm::RefVector approach, namely that the ScoutingPFJets and the ScoutingParticles are produced in the same module, making it impossible for one collection to contain refs to the other. Is this correct, and do I need to create a separate producer for ScoutingParticles in this approach?

—
Reply to this email directly or view it on GitHub<https://github.com/cms-sw/cmssw/pull/10378#issuecomment-129843381>.

fwyzard commented:
You can technically do everything in a single producer, look for
getRefBeforePut .

David, if you want to avoid using a RefVector that's fine by me, but I
would still ask to somehow store some kind of pointer (a RefProd ? its bare
collection id?) to the collection of constituents.

Ciao,
.Andrea
On 11 Aug 2015 13:21, "Dustin Anderson" <notifications@github.com> wrote:

> I think there is a technical issue with the edm::RefVector approach,
> namely that the ScoutingPFJets and the ScoutingParticles are produced in
> the same module, making it impossible for one collection to contain refs to
> the other. Is this correct, and do I need to create a separate producer for
> ScoutingParticles in this approach?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10378#issuecomment-129843381>.
>

davidlange6 commented:
> On Aug 11, 2015, at 1:36 PM, Andrea Bocci <notifications@github.com> wrote:
> 
> You can technically do everything in a single producer, look for
> getRefBeforePut .
> 
> David, if you want to avoid using a RefVector that's fine by me, but I
> would still ask to somehow store some kind of pointer (a RefProd ? its bare
> collection id?) to the collection of constituents.

Hi Andrea

the agreement we had in place did not include this - but we  can rediscuss this afternoon.


> 
> Ciao,
> .Andrea
> On 11 Aug 2015 13:21, "Dustin Anderson" <notifications@github.com> wrote:
> 
> > I think there is a technical issue with the edm::RefVector approach,
> > namely that the ScoutingPFJets and the ScoutingParticles are produced in
> > the same module, making it impossible for one collection to contain refs to
> > the other. Is this correct, and do I need to create a separate producer for
> > ScoutingParticles in this approach?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/cms-sw/cmssw/pull/10378#issuecomment-129843381>.
> >
> —
> Reply to this email directly or view it on GitHub.
>

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

fwyzard commented on HLTrigger/JetMET/plugins/HLTScoutingPFProducer.cc:
```diff
@@ -0,0 +1,249 @@
+                    int matchIndex = -1;
+                    int outIndex = 0;
+                    for(auto &outCand : *outPFCandidates){
+                        float dR = sqrt(pow(cand->eta() - outCand.eta(), 2) + pow(cand->phi() - outCand.phi(), 2));
```
trivial optimisation: instead of computing the square root for every check, you should compute 
```
const float minDR2 = minDR * minDR;
```
once, and compare it with
```
float dR2 = pow(cand->eta() - outCand.eta(), 2) + pow(cand->phi() - outCand.phi(), 2);
```

fwyzard commented on HLTrigger/JetMET/plugins/HLTScoutingPFProducer.cc:
```diff
@@ -0,0 +1,249 @@
+
+    private:
+        virtual void produce(edm::Event&, const edm::EventSetup&) override;
+
```
as a safety precaution, can you declare all the data members that are initialised in the constructor initialiser list as `const` ?

fwyzard commented on HLTrigger/JetMET/plugins/HLTScoutingPFProducer.cc:
```diff
@@ -0,0 +1,249 @@
+    doMet(iConfig.getParameter<bool>("doMet"))
+{
+    //register products
+    produces<ScoutingPFJetCollection>("scoutingPFJets");
```
just a remark: you could put these three collections into the event without an additional label

fwyzard commented on HLTrigger/JetMET/plugins/HLTScoutingPFProducer.cc:
```diff
@@ -0,0 +1,249 @@
+
+#include "DataFormats/Math/interface/deltaR.h"
+
+class HLTScoutingPFProducer : public edm::stream::EDProducer<> {
```
I did not try to do it, but I'm guessing this could safely be a `::global` producer instead of a `::stream` producer ?

fwyzard commented:
hi @davidlange6 ,
about the RefProd to the PF constituents: after re-reading the code of the producer, I agree we can do without it.
The collection of constituents is produced by the same producer as the pf jets, so the implicit association should be reasonably safe.

.A

davidlange6 commented on DataFormats/Scouting/interface/ScoutingVertex.h:
```diff
@@ -0,0 +1,32 @@
+        double x_;
+        double y_;
+        double z_;
+        double zError_;
```
@duanders  -was there a particular reason for these to be doubles instead of floats?

duanders commented:
Hi @fwyzard , @davidlange6 , 
I addressed all the comments you gave above.  The producers are changed to `global::EDProducers`, and the ScoutingVertex data members are now floats.

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

fwyzard commented:
hi Dustin, thanks for the changes

fwyzard commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

davidlange6 commented:
all - any last comments on this PR?

Martin-Grunewald commented:
I think one or both of the BuildFiles in HLTrigger/JetMET would need a 
+use DataFormats/Scouting or so, given taht both BuildFiles refer
to the other DataFormats packages used there...

IOW, did you test whether you can write and read back the new classes?

cmsbuild commented:
+1
Tested at: 217678ac5e17f413155c7ba1a50d557c5a08b60f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/7189/summary.html

duanders commented:
Hi @Martin-Grunewald ,
Yes -- I've tested that the producers work, and I can read back the class information using FWLite and with an example EDAnalyzer.  It doesn't seem to need the line in the BuildFile.  @davidlange6 @fwyzard any comment on this?  

One last-minute thing that I just caught: the accessor methods in the classes should be declared `const`.

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

Martin-Grunewald commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 3414813ac15caa4570cb4cf7f62877c0bb29933e
 When I ran the RelVals I found an error in the following worklfows: 
 25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

----- Begin Fatal Exception 14-Aug-2015 07:34:27 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Processing run: 1 lumi: 1 event: 1

   [1] Running path 'validation_step'

   [2] Calling event method for module MixingModule/'mix'

Exception Message:

RootInputFileSequence::readOneSpecified(): no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>
50202.0 step3
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step3_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

----- Begin Fatal Exception 14-Aug-2015 07:38:23 CEST-----------------------

An exception of category 'Configuration' occurred while

   [0] Processing run: 1 lumi: 1 event: 1

   [1] Running path 'validation_step'

   [2] Calling event method for module MixingModule/'mix'

Exception Message:

RootInputFileSequence::readOneSpecified(): no input files specified for secondary input source.

----- End Fatal Exception -------------------------------------------------

</pre>
4.53 step1
<pre>DAS Error</pre>
1000.0 step1
<pre>DAS Error</pre>
1003.0 step1
<pre>DAS Error</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/7257/summary.html

davidlange6 commented:
@duanders - yes, you should have needed to add DataFormats/Scouting to a BuildFile when you use it directly. I'm surprised it works without. Could you add it?

fwyzard commented:
I think it works because the data formats are just header files:
  - at compile time they are found anyway
  - at link time nothing else is needed
  - at run time the dictionaries are found by the plugin system (?)

Anyway it is a good idea to add it explicitly, otherwise it would break if part of the data formats are moved to a .cc file.

One last thing, though it's not a requirement, just a suggestion: could you do a `git rebase -i` and merge all the commits into a single one ?

Something like
```
cmsrel CMSSW_7_4_9
cd CMSSW_7_4_9/src
cmsenv
git cms-init
git checkout scoutingpackingcode74X
git rebase -i CMSSW_7_4_9
# change the first commit from 'pick' to 'r'
# change all other commits from  'pick' to 'f'
# save and exit the editor
# when prompted, give a new commit message, something like "implement data formats and producers for Scouting with jets"

# check that everything is fine, compile, test, etc.

git push -f
```

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

duanders commented:
@fwyzard , thanks for clarifying.  To be safe, I added the line.  

I'll try rebasing now to clean up the commits.

cmsbuild commented:
Pull request #10378 was updated. @Martin-Grunewald, @perrotta, @cmsbuild, @fwyzard can you please check and sign again.

Martin-Grunewald commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e13213cfa97347a2fe1eaae4a26c03bbcd09cce9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/7371/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10378/7371/summary.html

Martin-Grunewald commented:
+1
